src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -3349,6 +3449,330 @@ public <T> T invokeAny(Collection<? extends Callable<T>> tasks,
+     */
+    private DelayScheduler startDelayScheduler() {
+        DelayScheduler ds;
+        if ((ds = delayScheduler) == null) {
+            boolean start = false;
+            String name = poolName + "-delayScheduler";
+            if (workerNamePrefix == null)
+                asyncCommonPool();  // override common parallelism zero
+            lockRunState();
+            try {
+                if ((ds = delayScheduler) == null) {
+                    ds = delayScheduler = new DelayScheduler(this, name);
+                    start = true;
+                }
+            } finally {
+                unlockRunState();
+            }
+            if (start) { // start outside of lock
+                // exceptions on start passed to (external) callers
+                SharedThreadContainer ctr;
+                if ((ctr = container) != null)
+                    ctr.start(ds);
+                else
+                    ds.start();
+            }
+        }
+        return ds;
+    }
+
+    /**
+     * Callback upon starting DelayScheduler
+     */
+    final void onDelaySchedulerStart() {
+        WorkQueue q;           // set up default submission queue
+        if ((q = submissionQueue(0, false)) != null)
+            q.unlockPhase();
+    }
+
+    /**
+     * Arranges execution of a ready task from DelayScheduler, or
+     * cancels it on error
+     */
+    final void executeReadyScheduledTask(ScheduledForkJoinTask<?> task) {
+        if (task != null) {
+            WorkQueue q;
+            boolean cancel = false;
+            try {
+                if ((q = externalSubmissionQueue(false)) == null)
+                    cancel = true; // terminating
+                else
+                    q.push(task, this, false);
+            } catch(Error | RuntimeException ex) {
+                cancel = true;     // OOME or VM error
+            }
+            if (cancel)
+                task.trySetCancelled();
+        }
+    }
+
+    /**
+     * Arranges delayed execution of a ScheduledForkJoinTask via the
+     * DelayScheduler, creating and starting it if necessary.
+     * @return the task
+     */
+    final <T> ScheduledForkJoinTask<T> scheduleDelayedTask(ScheduledForkJoinTask<T> task) {
+        DelayScheduler ds;
+        if (((ds = delayScheduler) == null &&
+             (ds = startDelayScheduler()) == null) ||
+            task == null || (runState & SHUTDOWN) != 0L)
+            throw new RejectedExecutionException();
+        ds.pend(task);
+        return task;
+    }
+
+    /**
+     * Submits a one-shot task that becomes enabled after the given
+     * delay, At which point it will execute unless explicitly
+     * cancelled, or fail to execute (eventually reporting
+     * cancellation) when encountering resource exhaustion, or the
+     * pool is {@link #shutdownNow}, or is {@link #shutdown} when
+     * otherwise quiescent and {@link #cancelDelayedTasksOnShutdown}
+     * is in effect.
```

Comment:
In passing, I'm wondering if we should if the term "enabled" should be defined somewhere. SES uses it but it doesn't appear to define it.

Reply:
Added: 
 Delayed actions become <em>enabled</em> and behave as ordinary submitted tasks when their delays elapse.

Reply:
Thanks. Also here we have ".. after the given delay, At which" so need to smooth that out.

Reply:
OK. I also renamed executeReadyScheduledTask to executeEnabledScheduledTask and checked consistency of usages elsewhere.